pyflakes/checker.py

```diff
@@ -952,21 +968,27 @@ def IMPORT(self, node):
 
     def IMPORTFROM(self, node):
         if node.module == '__future__':
-            if not self.futuresAllowed:
+            # __future__ can only appear in module/doctest scope and
+            # should not have been disabled already in handleNode and
+            # the scope must not have any other type of binding.
+            if (not isinstance(self.scope, ModuleScope) or
+                    not self.scope._futures_allowed or
+                    any(not isinstance(binding, FutureImportation)
+                        for binding in self.scope.values())):
+                self.scope._futures_allowed = False
```

Comment:
I guess that makes sense, but if the code requires some explanation, then I'd like to improve the code so future readers won't run into the same difficulties I'm having understanding the code now.

What can be done to make this code more digestible? Can we:

  - break up the compound conditional expressions into several `if` statements?
  - assign intermediate names to some of the subexpressions?
  - extract some of the logic to well-named private methods or functions?
  - consolidate the logic?